controller/lib/core/blower_fsm.cpp

```diff
@@ -0,0 +1,187 @@
+/* Copyright 2020, RespiraWorks
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "blower_fsm.h"
+
+#include "hal.h"
+#include "new.h"
+#include "units.h"
+
+// This module is responsible for generating an ideal pressure graph over time.
+// It's then up to the PID module to realize this pressure, as closely as it
+// can.
+//
+// The main() loop queries blower_fsm_get_setpoint() on each iteration.
+// get_setpoint() delegates to a "breath FSM" (e.g. PressureControlFsm), which
+// is responsible for *one breath* with a fixed mode and VentParams.  When that
+// breath ends, we create a new inner FSM with (potentially) new params.
+//
+// Decision that params should only change at breath boundaries:
+// https://respiraworks.slack.com/archives/CV4MTUJHF/p1588001011133500
+
+namespace {
+
+// A "breath finite state machine" where the blower is always off.
+//
+// All breath FSMs should implement the following "duck-typed API".
+//
+//  - <constructor>(const VentParams& params): Constructs a new FSM for a
+//    single breath, with the given params.  Those params don't change during
+//    the life of the FSM.
+//
+//  - Pressure get_setpoint(): Gets the pressure that the fan should be trying
+//    to hit at this point in time.
+//
+//  - bool finished(): Has this breath FSM completed its work (namely, running
+//    a single breath)?  If so, it is ready to be replaced with a new one.
+//
+class OffFsm {
+public:
+  OffFsm() = default;
+  explicit OffFsm(const VentParams &) {}
+  Pressure get_setpoint() { return kPa(0); }
+  bool finished() { return true; }
+};
+
+// "Breath finite state machine" for pressure control mode.
+//
+// Currently our pressure-control mode drives a simple square wave: We go from
+// PIP pressure on inhale to PEEP pressure on exhale.
+class PressureControlFsm {
+public:
+  explicit PressureControlFsm(const VentParams &params)
+      : inspire_pressure_(cmH2O(params.pip_cm_h2o)),
+        expire_pressure_(cmH2O(params.peep_cm_h2o)),
+        inspire_end_(Hal.now() + time_per_inspire(params)),
```

Comment:
Can this compute Hal.now() only once, for peace of mind?

Reply:
Done.  Means I can't use an initialization list (sad), but okay, we deal.